stop_and_wait_plugin/src/stop_and_wait_plugin.cpp

```diff
@@ -0,0 +1,386 @@
+    }
+
+    std::vector<cav_msgs::TrajectoryPlanPoint> StopandWait::compose_trajectory_from_centerline(
+    const std::vector<PointSpeedPair>& points, const cav_msgs::VehicleState& state)
+    {        
+        int nearest_pt_index = getNearestPointIndex(points,state);
+        std::vector<PointSpeedPair> future_points(points.begin() + nearest_pt_index + 1, points.end()); // Points in front of current vehicle position
+
+        //Get yaw - geometrically
+        std::vector<float> yaw_values;
+        for(size_t i=0 ;i < future_points.size()-1 ;i++)
+        {
+            float yaw = atan((future_points[i+1].point.y() - future_points[i].point.y())/ (future_points[i+1].point.x() - future_points[i].point.x()));
+            yaw_values.push_back(yaw);
+        }
+        yaw_values.push_back(0.0); //No rotation from last point
+
+        //get target time from speed
+        std::vector<double> target_times;
+        std::vector<lanelet::BasicPoint2d> trajectory_locations;
+        std::vector<double> trajectory_speeds;
+        //split point speed pair
+        splitPointSpeedPairs(future_points,&trajectory_locations,&trajectory_speeds);
+        std::vector<double> downtracks = carma_wm::geometry::compute_arc_lengths(trajectory_locations);
+
+        //get trajectory time from distance and speed
+        speed_to_time(downtracks, trajectory_speeds, &target_times, jerk_);
+        std::vector <cav_msgs::TrajectoryPlanPoint> traj;
+        ros::Time start_time = ros::Time::now();
+        cav_msgs::TrajectoryPlanPoint traj_prev;
+        traj_prev.x= future_points[0].point.x();
+        traj_prev.y = future_points[0].point.y();
+        traj_prev.yaw = yaw_values[0];
+        traj_prev.target_time = start_time;
+        double delta_time = maneuver_time_/future_points.size();
+
+        for (size_t i=0; i < future_points.size(); i++)
+        {
+            cav_msgs::TrajectoryPlanPoint traj_point;
+            if(trajectory_speeds[i] > 0){
+                traj_point.x = future_points[i].point.x();
+                traj_point.y = future_points[i].point.y();
+                traj_point.yaw = yaw_values[i];
+                traj_point.target_time = start_time + ros::Duration(target_times[i]);
+            }
+            else    //speed_to_time doesn't work for 0.0 speed
+            {
+                traj_point.x=traj_prev.x;
+                traj_point.y=traj_prev.y;
+                traj_point.yaw=traj_prev.yaw;
+                traj_point.target_time = traj_prev.target_time + ros::Duration(0.1);
+                
+            }
+            traj_point.controller_plugin_name = "default";
+            traj_point.planner_plugin_name =plugin_discovery_msg_.name;
+            traj.push_back(traj_point);
+            traj_prev = traj_point;
+        }
+        
+        return traj;
+    }
+
+    void StopandWait::speed_to_time(const std::vector<double>& downtrack, const std::vector<double>& speeds, std::vector<double>* times, double jerk)
```

Comment:
Looking at this now, I'd just recommend making times a reference to avoid using raw pointers. Should accomplish the same results. And I'd make sure the doc comments mention that times will be modified as a result. Either that, or if it's imperative that times be an empty vector prior to this function, i'd say just make it the return value of this function instead of an input.

Reply:
Generally, I'd favor return values almost all the time unless it's expected that the object being returned will be very large in size (and even then, I'd likely want to see performance profiling data indicating that we'd actually have a problem before optimizing like that). I just think that's easier to follow most of the time as it better highlights where data transformations are happening and how.